autode/wrappers/NWChem.py

```diff
@@ -1,133 +1,194 @@
+            print_constraints(inp_file, molecule)
 
-            if calc.distance_constraints or calc.cartesian_constraints:
-                force_constant = 20
-                print('constraints', file=inp_file)
-                if calc.distance_constraints:
-                    for atom_ids in calc.distance_constraints.keys():  # nwchem counts from 1 so increment atom ids by 1
-                        dist = calc.distance_constraints[atom_ids] / Constants.a02ang       # Constraints are in Bohr
-                        print(f'  spring bond {atom_ids[0] + 1} {atom_ids[1] + 1} {force_constant} {dist:.3f}' + str(
-                            atom_ids[0] + 1), file=inp_file)
-
-                if calc.cartesian_constraints:
-                    constrained_atoms = [i + 1 for i in calc.cartesian_constraints]
-                    list_of_ranges = []
-                    used_atoms = []
-                    for atom in constrained_atoms:
-                        rang = []
-                        if atom not in used_atoms:
-                            while atom in constrained_atoms:
-                                used_atoms.append(atom)
-                                rang.append(atom)
-                                atom += 1
-                            if len(rang) in (1, 2):
-                                list_of_ranges += rang
-                            else:
-                                range_string = str(rang[0]) + ':' + str(rang[-1])
-                                list_of_ranges.append(range_string)
-                    print('  fix atom', end=' ', file=inp_file)
-                    print(*list_of_ranges, sep=' ', file=inp_file)
-                print('end', file=inp_file)
-
-            if calc.point_charges:
+            if calc.input.point_charges is not None:
                 print('bq')
                 for charge, x, y, z in calc.point_charges:
-                    print(f'{x:^12.8f} {y:^12.8f} {z:^12.8f} {charge:^12.8f}', file=inp_file)
+                    print(f'{x:^12.8f} {y:^12.8f} {z:^12.8f} {charge:^12.8f}',
+                          file=inp_file)
                 print('end')
 
             print(f'memory {Config.max_core} mb', file=inp_file)
 
             print(*keywords, sep='\n', file=inp_file)
 
-            # Will used partial an ESP initialisation to generate partial atomic charges - more accurate than
-            # the standard Mulliken analysis (or at least less sensitive to the method)
+            # Will used partial an ESP initialisation to generate partial
+            # atomic charges - more accurate than the standard Mulliken
+            # analysis (or at least less sensitive to the method)
             print('task esp', file=inp_file)
 
         return None
 
+    def get_input_filename(self, calc):
+        return f'{calc.name}_nwchem.nw'
+
+    def get_output_filename(self, calc):
+        return f'{calc.name}_nwchem.out'
+
+    def execute(self, calc):
```

Comment:
I'm not wild about having a self.execute() and exectute() function in the same file

Reply:
Fair point. Suggestions?

Reply:
setup_execution() and execute_calc()?
although setup_execution sounds a bit dark

Reply:
gone with run_external() in base - think it's the most clear